cpp/src/ffi_go/ir/deserializer.cpp

```diff
@@ -231,10 +209,88 @@ auto deserialize_wildcard_match(
         log_event->m_log_message.m_data = deserializer->m_log_event.m_log_message.data();
         log_event->m_log_message.m_size = deserializer->m_log_event.m_log_message.size();
         log_event->m_timestamp = deserializer->m_timestamp;
+        log_event->m_utc_offset = static_cast<epoch_time_ms_t>(deserializer->m_utc_offset.count());
         *matching_query = matching_query_idx;
         return static_cast<int>(IRErrorCode::IRErrorCode_Success);
     }
 }
+
+auto deserialize_all_utc_offset_change_packets(
+        BufferReader& ir_buf,
+        encoded_tag_t& tag,
+        clp::UtcOffset& utc_offset
+) -> IRErrorCode {
+    while (UtcOffsetChange == tag) {
+        if (auto const err{clp::ffi::ir_stream::deserialize_utc_offset_change(ir_buf, utc_offset)})
+        {
+            return err;
+        }
+        if (auto const err{deserialize_tag(ir_buf, tag)}) {
+            return err;
+        }
+    }
+    return IRErrorCode::IRErrorCode_Success;
+}
+
+template <class encoded_variable_t>
+auto deserialize_to_next_log_event(BufferReader& ir_buf, Deserializer* deserializer)
+        -> IRErrorCode {
+    static_assert(
+            (std::is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>
+             || std::is_same_v<encoded_variable_t, four_byte_encoded_variable_t>)
+    );
+
+    encoded_tag_t tag{};
+    epoch_time_ms_t timestamp{};
+    IRErrorCode err{};
+    std::optional<clp::UtcOffset> utc_offset;
```

Comment:
Using `std::optional` for deserialized utc offsets, and set it only when we return success. This will ensure the state of deserializer is unchanged if an error happens.

Reply:
Since it is possible to fail outside of `deserialize_to_next_log_event` (even when it returns success), this isn't actually a complete defense... We would need to only update the stored utc_offset after we decide to return an updated ir position.